scipy/stats/_fit.py

```diff
@@ -164,10 +198,99 @@ def plot(self, ax=None):
+
+        if qq:
+            qp = "Quantiles"
+            plot_type = 'Q-Q'
+            x = self._dist.ppf(ps, *fit_params)
+            y = data
+        else:
+            qp = "Percentiles"
+            plot_type = 'P-P'
+            x = ps
+            y = self._dist.cdf(data, *fit_params)
+
+        ax.plot(x, y, '.', label=f'Fitted Distribution {plot_type}',
+                color='C0', zorder=1)
+        xlim = ax.get_xlim()
+        ylim = ax.get_ylim()
+        lim = [min(xlim[0], ylim[0]), max(xlim[1], ylim[1])]
+        if not qq:
+            lim = max(lim[0], 0), min(lim[1], 1)
+
+        if self.discrete and qq:
+            q_min, q_max = int(lim[0]), int(lim[1]+1)
+            q_ideal = np.arange(q_min, q_max)
+            # q_ideal = np.unique(self._dist.ppf(ps, *fit_params))
+            ax.plot(q_ideal, q_ideal, 'o', label='Reference', color='k',
+                    alpha=0.25, markerfacecolor='none', clip_on=True)
+        elif self.discrete and not qq:
+            # The intent of this is to match the plot that would be produced
+            # if x were continuous on [0, 1] and y were cdf(ppf(x)).
+            # It can be approximated by letting x = np.linspace(0, 1, 1000),
+            # but this might not look great when zooming in. The vertical
+            # portions are included to indicate where the transition occurs
+            # where the data completely obscures the horizontal portions.
+            p_min, p_max = lim
+            a, b = self._dist.support(*fit_params)
+            p_min = max(p_min, 0 if np.isfinite(a) else 1e-3)
+            p_max = min(p_max, 1 if np.isfinite(b) else 1-1e-3)
+            q_min, q_max = self._dist.ppf([p_min, p_max], *fit_params)
+            qs = np.arange(q_min-1, q_max+1)
+            ps = self._dist.cdf(qs, *fit_params)
+            ax.step(ps, ps, '-', label='Reference', color='k', alpha=0.25,
+                    clip_on=True)
+        else:
+            ax.plot(lim, lim, '-', label='Reference', color='k', alpha=0.25,
+                    clip_on=True)
+
+        ax.set_xlim(lim)
+        ax.set_ylim(lim)
+        ax.set_xlabel(rf"Fitted $\tt {self._dist.name}$ Theoretical {qp}")
+        ax.set_ylabel(f"Data {qp}")
+        ax.set_title(rf"Fitted $\tt {self._dist.name}$ {plot_type} Plot")
+        ax.legend(*ax.get_legend_handles_labels())
+        ax.set_aspect('equal')
+        return ax
+
+    def _qq_plot(self, **kwargs):
+        return self._qp_plot(qq=True, **kwargs)
+
+    def _pp_plot(self, **kwargs):
+        return self._qp_plot(qq=False, **kwargs)
+
+    def _plotting_positions(self, n, a=.5):
+        # See https://en.wikipedia.org/wiki/Q%E2%80%93Q_plot#Plotting_positions
+        k = np.arange(1, n+1)
+        return (k-a) / (n + 1 - 2*a)
+
+    def _cdf_plot(self, ax, fit_params):
+        data = np.sort(self._data)
+        ecdf = self._plotting_positions(len(self._data))
+        ls = '--' if self.discrete else '-'
```

Comment:
Perhaps this is more what we want:

```suggestion
ls = '--' if len(np.unique(data)) < 30 else '.'
```
The thought is that when we have too many steps, the dashed line starts looking funny. The number of steps is the number of unique elements in `data`. It so happens that my `nbinom` example had fewer discrete steps, but it wasn't inherently because it was discrete. 

With this suggestion, here is data from a discrete distribution with lots of steps:
![image](https://user-images.githubusercontent.com/6570539/189276365-fb48393f-3d8d-4d5e-803f-c185deb4aab5.png)

And data from a continuous distribution with few steps.
![image](https://user-images.githubusercontent.com/6570539/189276479-43156db2-f1c3-4384-8ff6-f9bc300d6565.png)

Reply:
My concern is about differentiating the lines without colours (for colour deficiency and BW printing). But actually we can just use markers. Because yes I agree that otherwise it does not look very good and I would also prefer not to add any custom logic there. In the end plot styling is very much personal and users can do that themselves as we return the mpl object.

So here we need to use different markers and different step size (`markevery`) https://matplotlib.org/stable/gallery/lines_bars_and_markers/markevery_demo.html

Reply:
Using `.` as the style already does what we want.

Reply:
Are you saying then we can just have `ls = '.' if self.discrete else '-'`?

Reply:
Close. But the `.` needs to go on the right, and `-` should be `--` and on the left. 
```suggestion
        ls = '--' if self.discrete else '.'
```
But I still think the other suggestion that looks at the number of unique elements is better for a default. I wish I had done that initially.

Reply:
My opinion is not that strong here. Let's go with the first suggestion and see how it goes. Plotting is fine and we are pretty much free to change things there I think.